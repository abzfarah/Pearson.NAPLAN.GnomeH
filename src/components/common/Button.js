
import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';

import CSSClassnames from '../utils/CSSClassnames';

const CLASS_ROOT = CSSClassnames.BUTTON;

export default class Button extends Component {
  constructor () {
    super();
    this.state = {
      mouseActive: false,
      focus: false
    };
  }

  render () {
    const {
      a11yTitle, accent, align, children, className, fill, href, icon, id,
      label, onClick, onBlur, onFocus, onMouseDown, onMouseUp, plain, primary,
      secondary, type, ...props
    } = this.props;

    const buttonPlain = (plain !== undefined ? plain : (icon && ! label));

    let buttonIcon;
    if (icon) buttonIcon =
      <span className={`${CLASS_ROOT}__icon`}>{icon}</span>;

    let hasIcon = buttonIcon !== undefined;
    let buttonChildren = React.Children.map(children, child => {
      if (child && child.type && child.type.icon) {
        hasIcon = true;
        child = <span className={`${CLASS_ROOT}__icon`}>{child}</span>;
      }
      return child;
    });

    const classes = classnames(
      CLASS_ROOT,
      {
        [`${CLASS_ROOT}--focus`]: this.state.focus,
        [`${CLASS_ROOT}--primary`]: primary,
        [`${CLASS_ROOT}--secondary`]: secondary,
        [`${CLASS_ROOT}--accent`]: accent,
        [`${CLASS_ROOT}--disabled`]: !onClick && !href,
        [`${CLASS_ROOT}--fill`]: fill,
        [`${CLASS_ROOT}--plain`]: buttonPlain,
        [`${CLASS_ROOT}--icon`]: icon || hasIcon,
        [`${CLASS_ROOT}--align-${align}`]: align
      },
      className
    );

    if (!buttonChildren) {
      buttonChildren = label;
    }

    const Tag = href ? 'a' : 'button';
    let buttonType;
    if (!href) {
      buttonType = type;
    }

    return (
      <Tag {...props} href={href} id={id} type={buttonType}
        className={classes} aria-label={a11yTitle}
        onClick={onClick}
        disabled={!onClick && !href}
        onMouseDown={(event) => {
          this.setState({ mouseActive: true });
          if (onMouseDown) {
            onMouseDown(event);
          }
        }}
        onMouseUp={(event) => {
          this.setState({ mouseActive: false });
          if (onMouseUp) {
            onMouseUp(event);
          }
        }}
        onFocus={(event) => {
          if (this.state.mouseActive === false) {
            this.setState({ focus: true });
          }
          if (onFocus) {
            onFocus(event);
          }
        }}
        onBlur={(event) => {
          this.setState({ focus: false });
          if (onBlur) {
            onBlur(event);
          }
        }}>
        {buttonIcon}
        {buttonChildren}
      </Tag>
    );
  }
};

Button.propTypes = {
  a11yTitle: PropTypes.string,
  accent: PropTypes.bool,
  align: PropTypes.oneOf(['start', 'center', 'end']),
  fill: PropTypes.bool,
  href: PropTypes.string,
  icon: PropTypes.element,
  id: PropTypes.string,
  label: PropTypes.node,
  onClick: PropTypes.func,
  plain: PropTypes.bool,
  primary: PropTypes.bool,
  secondary: PropTypes.bool,
  type: PropTypes.oneOf(['button', 'reset', 'submit'])
};

Button.defaultProps = {
  type: 'button'
};
