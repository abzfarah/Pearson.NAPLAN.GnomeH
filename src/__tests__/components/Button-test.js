import React from 'react';
import renderer from 'react/lib/ReactTestRenderer';

import Button from '../../components/common/Button';

import FakeIcon from '../../actions/FakeIcon';

describe('Button', () => {
  it('has correct default options', () => {
    const component = renderer.create(
      <Button label='Test me' />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has correct primary={true} rendering', () => {
    const component = renderer.create(
      <Button label='Test me' primary={true} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has correct plain={true} rendering', () => {
    const component = renderer.create(
      <Button label='Test me' plain={true} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has correct icon rendering', () => {
    const component = renderer.create(
      <Button icon={<FakeIcon />} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has correct icon as children rendering', () => {
    const component = renderer.create(
      <Button><FakeIcon /></Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has text as children rendering', () => {
    const component = renderer.create(
      <Button>test</Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has className property rendering', () => {
    const component = renderer.create(
      <Button className="test">test</Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has microdata properties rendering', () => {
    const component = renderer.create(
      <Button itemScope={true} itemType="http://schema.org/Article"
        itemProp="test">test</Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
