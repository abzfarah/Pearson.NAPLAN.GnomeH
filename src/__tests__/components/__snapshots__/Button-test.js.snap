exports[`Button has className property rendering 1`] = `
<button
  aria-label={undefined}
  className="grommetux-button grommetux-button--disabled test"
  disabled={true}
  href={undefined}
  id={undefined}
  onBlur={[Function onBlur]}
  onClick={undefined}
  onFocus={[Function onFocus]}
  onMouseDown={[Function onMouseDown]}
  onMouseUp={[Function onMouseUp]}
  type="button">
  test
</button>
`;

exports[`Button has correct default options 1`] = `
<button
  aria-label={undefined}
  className="grommetux-button grommetux-button--disabled"
  disabled={true}
  href={undefined}
  id={undefined}
  onBlur={[Function onBlur]}
  onClick={undefined}
  onFocus={[Function onFocus]}
  onMouseDown={[Function onMouseDown]}
  onMouseUp={[Function onMouseUp]}
  type="button">
  Test me
</button>
`;

exports[`Button has correct icon as children rendering 1`] = `
<button
  aria-label={undefined}
  className="grommetux-button grommetux-button--disabled grommetux-button--icon"
  disabled={true}
  href={undefined}
  id={undefined}
  onBlur={[Function onBlur]}
  onClick={undefined}
  onFocus={[Function onFocus]}
  onMouseDown={[Function onMouseDown]}
  onMouseUp={[Function onMouseUp]}
  type="button">
  <span
    className="grommetux-button__icon">
    <svg
      height="24px"
      role="img"
      version="1.1"
      viewBox="0 0 24 24.3616"
      width="24px" />
  </span>
</button>
`;

exports[`Button has correct icon rendering 1`] = `
<button
  aria-label={undefined}
  className="grommetux-button grommetux-button--disabled grommetux-button--plain grommetux-button--icon"
  disabled={true}
  href={undefined}
  id={undefined}
  onBlur={[Function onBlur]}
  onClick={undefined}
  onFocus={[Function onFocus]}
  onMouseDown={[Function onMouseDown]}
  onMouseUp={[Function onMouseUp]}
  type="button">
  <span
    className="grommetux-button__icon">
    <svg
      height="24px"
      role="img"
      version="1.1"
      viewBox="0 0 24 24.3616"
      width="24px" />
  </span>
</button>
`;

exports[`Button has correct plain={true} rendering 1`] = `
<button
  aria-label={undefined}
  className="grommetux-button grommetux-button--disabled grommetux-button--plain"
  disabled={true}
  href={undefined}
  id={undefined}
  onBlur={[Function onBlur]}
  onClick={undefined}
  onFocus={[Function onFocus]}
  onMouseDown={[Function onMouseDown]}
  onMouseUp={[Function onMouseUp]}
  type="button">
  Test me
</button>
`;

exports[`Button has correct primary={true} rendering 1`] = `
<button
  aria-label={undefined}
  className="grommetux-button grommetux-button--primary grommetux-button--disabled"
  disabled={true}
  href={undefined}
  id={undefined}
  onBlur={[Function onBlur]}
  onClick={undefined}
  onFocus={[Function onFocus]}
  onMouseDown={[Function onMouseDown]}
  onMouseUp={[Function onMouseUp]}
  type="button">
  Test me
</button>
`;

exports[`Button has microdata properties rendering 1`] = `
<button
  aria-label={undefined}
  className="grommetux-button grommetux-button--disabled"
  disabled={true}
  href={undefined}
  id={undefined}
  itemProp="test"
  itemScope={true}
  itemType="http://schema.org/Article"
  onBlur={[Function onBlur]}
  onClick={undefined}
  onFocus={[Function onFocus]}
  onMouseDown={[Function onMouseDown]}
  onMouseUp={[Function onMouseUp]}
  type="button">
  test
</button>
`;

exports[`Button has text as children rendering 1`] = `
<button
  aria-label={undefined}
  className="grommetux-button grommetux-button--disabled"
  disabled={true}
  href={undefined}
  id={undefined}
  onBlur={[Function onBlur]}
  onClick={undefined}
  onFocus={[Function onFocus]}
  onMouseDown={[Function onMouseDown]}
  onMouseUp={[Function onMouseUp]}
  type="button">
  test
</button>
`;
